ftp/src/main/scala/akka/stream/alpakka/ftp/model.scala

```diff
@@ -60,45 +61,81 @@ sealed abstract class FtpFileSettings extends RemoteFileSettings {
  */
-final case class FtpSettings(
-    host: InetAddress,
-    port: Int = FtpSettings.DefaultFtpPort,
-    credentials: FtpCredentials = AnonFtpCredentials,
-    binary: Boolean = false,
-    passiveMode: Boolean = false,
-    configureConnection: FTPClient => Unit = _ => {}
+final class FtpSettings private (
+    val host: java.net.InetAddress,
+    val port: Int,
+    val credentials: FtpCredentials,
+    val binary: Boolean,
+    val passiveMode: Boolean,
+    val configureConnection: org.apache.commons.net.ftp.FTPClient => Unit
 ) extends FtpFileSettings {
-  def withPort(port: Int): FtpSettings =
-    copy(port = port)
 
-  def withCredentials(credentials: FtpCredentials): FtpSettings =
-    copy(credentials = credentials)
+  def withHost(value: java.net.InetAddress): FtpSettings = copy(host = value)
+  def withPort(value: Int): FtpSettings = copy(port = value)
+  def withCredentials(value: FtpCredentials): FtpSettings = copy(credentials = value)
+  def withBinary(value: Boolean): FtpSettings = if (binary == value) this else copy(binary = value)
+  def withPassiveMode(value: Boolean): FtpSettings = if (passiveMode == value) this else copy(passiveMode = value)
 
-  def withBinary(binary: Boolean): FtpSettings =
-    copy(binary = binary)
-
-  def withPassiveMode(passiveMode: Boolean): FtpSettings =
-    copy(passiveMode = passiveMode)
-
-  def withConfigureConnection(configureConnection: FTPClient => Unit): FtpSettings =
-    copy(configureConnection = configureConnection)
+  /**
+   * Scala API:
+   * Sets the configure connection callback.
+   */
+  def withConfigureConnection(value: FTPClient => Unit): FtpSettings =
+    copy(configureConnection = value)
 
   /**
    * Java API:
    * Sets the configure connection callback.
    */
   def withConfigureConnectionConsumer(configureConnection: java.util.function.Consumer[FTPClient]): FtpSettings =
     copy(configureConnection = configureConnection.accept)
+
+  private def copy(
+      host: java.net.InetAddress = host,
+      port: Int = port,
+      credentials: FtpCredentials = credentials,
+      binary: Boolean = binary,
+      passiveMode: Boolean = passiveMode,
+      configureConnection: FTPClient => Unit = configureConnection
+  ): FtpSettings = new FtpSettings(
+    host = host,
+    port = port,
+    credentials = credentials,
+    binary = binary,
+    passiveMode = passiveMode,
+    configureConnection = configureConnection
+  )
+
+  override def toString =
+    s"""FtpSettings(host=$host,port=$port,credentials=$credentials,binary=$binary,passiveMode=$passiveMode,configureConnection=$configureConnection)"""
 }
 
+/**
+ * FTP settings factory
+ */
 object FtpSettings {
 
-  /** Default FTP port */
+  /** Default FTP port (21) */
   final val DefaultFtpPort = 21
 
+  /** Scala API */
+  def apply(
+      host: java.net.InetAddress
+  ): FtpSettings = new FtpSettings(
+    host,
+    port = DefaultFtpPort,
```

Comment:
We have default parameters in the private constructor in the reference connector:

https://github.com/akka/alpakka/blob/909cdd29ad39ad6d8f85dd32c2534adcfcfecfa1/reference/src/main/scala/akka/stream/alpakka/reference/model.scala#L81-L84

Reply:
We should change that, as it could easily lead to buggy implementations of the manually-written copy method.